File: python/test_api_calls.py
#!/usr/bin/env python
"""
 Copyright (C) 2013 Digium, Inc.

 Erin Spiceland <espiceland@digium.com>

 See http://www.asterisk.org for more information about
 the Asterisk project. Please do not directly contact
 any of the maintainers of this project for assistance;
 the project provides a web site, mailing lists and IRC
 channels for your use.

 This program is free software, distributed under the terms of
 the GNU General Public License Version 2. See the LICENSE file
 at the top of the source tree.

"""

import sys
import os
import glob
import json
import requests
from pprint import pprint


HOST = '192.168.1.124'
PORT = '8088'
PATH = '/home/erin/asterisk_rest_libraries'


def main(argv):
    os.chdir(PATH)
    for jsonfile in glob.glob("*.json"):
        res = get_JSON_from_file(jsonfile)
        if res is None:
            continue

        for api in res['apis']:
            try:
                for op in api['operations']:
                    try:
                        uri = "http://%s:%s/stasis%s" \
                            % (HOST, PORT, api['path'])
                    except KeyError, e:
                        print "KeyError 1 in %s: %s" % (jsonfile, e)
                        continue

                    try:
                        paramObj = generate_params(op['parameters'])
                    except:
                        paramObj = {}

                    try:
                        if op['httpMethod'] == 'GET':
                            resp = requests.get(uri, params=paramObj)
                        elif op['httpMethod'] == 'POST':
                            resp = requests.post(uri, params=paramObj)
                        elif op['httpMethod'] == 'DELETE':
                            resp = requests.delete(uri, params=paramObj)
                        elif op['httpMethod'] == 'PUT':
                            resp = requests.put(uri, params=paramObj)
                    except requests.exceptions.ConnectionError:
                        print "Connection refused"
                        return 1

                    if resp is None:
                        continue

                    print "%s %s %s" \
                        % (op['responseClass'], op['httpMethod'], uri)

                    if resp.text:
                        print "\t%s: %s" % (resp.status_code, resp.text)
            except KeyError, e:
                print "KeyError 2 in %s: %s" % (jsonfile, e)
                continue


def get_JSON_from_file(jsonfile):
    f = open(PATH + "/" + jsonfile, 'r')
    jsonString = f.read()
    f.close()
    try:
        res = json.loads(jsonString)
    except Exception, e:
        print e
        pass

    return res


def generate_params(ops):
    res = {}
    for op in ops:
        if op['dataType'] == 'string':
            if op['allowMultiple']:
                res[op['name']] = ['example', 'string']
            else:
                res[op['name']] = 'example'
        elif op['dataType'] == 'number':
            if op['allowMultiple']:
                res[op['name']] = [20, 8]
            else:
                res[op['name']] = 20
        elif op['dataType'] == 'boolean':
            if op['allowMultiple']:
                res[op['name']] = [True, False]
            else:
                res[op['name']] = True

    return res

if __name__ == "__main__":
    sys.exit(main(sys.argv) or 0)

File: perl/test_api_calls.pl
#!/usr/bin/perl

use strict;
use LWP::UserAgent;
use Data::Dumper;
use JSON;

my $HOST = '192.168.1.124';
#my $HOST = '10.24.67.73';
my $PORT = '8088';
my $REQ = $ARGV[0];
my $JSON_PATH = '/home/erin/api_stuff';

my $ua = LWP::UserAgent->new;
my $j = JSON->new->utf8;

foreach my $file (glob("$JSON_PATH/*.json")) {
	# skip if file size is 0
	next if -z $file;
	my $res = get_JSON_from_file($file);

	# skip if JSON is missing some key things
	next if !$res->{'apis'};
	next if !$res->{'resourcePath'};

	foreach my $api (@{$res->{'apis'}}) {
		# skip if there are no operations
		next if !$api->{'operations'};
		foreach my $op (@{$api->{'operations'}}) {
			my $uri = "http://$HOST:$PORT/stasis" . $api->{'path'};
			if ($uri =~ /[{}]/) {
				print "\tSkipping object-specific request\n";
				next;
			}
			my $paramObj = generate_params($op->{'parameters'});
			my $response;
			if ($op->{'httpMethod'} eq 'GET') {
				my $paramString = make_param_string($paramObj);
				$response = $ua->get("$uri$paramString");
			} elsif ($op->{'httpMethod'} eq 'POST') {
				$response = $ua->post($uri, $paramObj);
			#} elsif ($op->{'httpMethod'} eq 'DELETE') {
			#	my $paramString = make_param_string($paramObj);
			#	$response = $ua->delete("$uri$paramString");
			} elsif ($op->{'httpMethod'} eq 'PUT') {
				$response = $ua->put($uri, $paramObj);
			}
			next if !$response;
			print $op->{'responseClass'}, ' ', $op->{'httpMethod'}, ' ',
				$uri, "\n";
			if ($response->content =~ /connection refused/i) {
				print "Can't contact server.\n";
				exit(1);
			}
			if ($response->content) {
				if (eval { JSON::XS::decode_json($response->content); 1; } ) {
					my $content = JSON::XS::decode_json($response->content);
					print "\t", $response->status_line, "\n";
				} else {
					print "\tINVALID JSON\n";
					print $response->content, "\n";
				}
			}
		}
	}
}

sub ddump($) {
	my $ref = shift;
	if ($ref) {
		print Data::Dumper::Dumper($ref), "\n";
	}
}

sub get_JSON_from_file($) {
	my $file = shift;
	open INF, $file;
	my $json = join('', <INF>);
	close INF;
	return JSON::XS::decode_json($json);
}

sub generate_params($) {
	my $params = shift;
	my $res = {};
	foreach my $param (@$params) {
		if ($param->{'dataType'} eq 'string') {
			$res->{$param->{'name'}} = $param->{'allowMultiple'}
				? 'example' : ['example', 'string'];
		} elsif ($param->{'dataType'} eq 'number') {
			$res->{$param->{'name'}} = $param->{'allowMultiple'}
				? 20 : ['20', '8'];
		} elsif ($param->{'dataType'} eq 'boolean') {
			$res->{$param->{'name'}} = $param->{'allowMultiple'}
				? 'true' : ['true', 'false'];
		}
	}
	return $res;
}

sub make_param_string($) {
	my $params = shift;
	my $strings = [];
	foreach my $name (keys %$params) {
		if (ref $params->{$name} eq 'ARRAY') {
			push @$strings, join('=', $name, join(',', @{$params->{$name}}));
		} elsif (ref $params->{$name} eq 'SCALAR') {
			push @$strings, join('=', $name, $params->{$name});
		}
	}
	return '?' . join('&', @$strings);
}

